src/Analyser/MutatingScope.php

```diff
@@ -2274,19 +2274,19 @@ private function resolveEqualType(Type $leftType, Type $rightType): BooleanType
 	 */
 	private function resolveConstantArrayTypeComparison(ConstantArrayType $leftType, ConstantArrayType $rightType, callable $valueComparisonCallback): BooleanType
 	{
-		$leftValueTypes = $leftType->getValueTypes();
+		if ($leftType->hasOptionalKeys() || $rightType->hasOptionalKeys()) {
+			return new BooleanType();
```

Comment:
We can tell when it's always false even when optional keys are involved. This code doesn't seem right.

if I have `array{a?: int}` and `array{b? int}`, it can be true because both arrays can be empty.

But if I have `array{c, int, d?: int}`, `array{e: int, f?: int}`, we can be sure it's always false.

Reply:
That is already handled by the supertype check in `resolveIdenticalType`. I've added these as test cases.